### DISCUSSION TITLE
Set or Unique technigues in CUE

### DISCUSSION THREAD

**Discussion by ekarlso** [cue-lang/cue#1213] [Q&A] [answered]
Assuming I have a list of def's or even numbers how can I filter it to a unique set of items maybe into a new list.

```
// I'd like to only get 1.2.3.4 but unsure if I can do a "seen" like technique to do it in CUE
test: [1,2,3,3,4,4]
```

**Comment by jlongtine** [comment, marked as answer]
@ekarlso I usually do something like this:

```
test: [1,2,3,3,4,4]
let tmp = { for i in test { "\(i)": i } }
test_uniq: [  for i, j in tmp { j } ]
```

**Reply by jlongtine** [reply]
You can play with it here: https://tip.cuelang.org/play/?id=7-C_lxLLbTb#cue@export@cue

**Reply by myitcv** [reply]
Noting also the `list` builtins the exist to help validate/transform lists: 

https://pkg.go.dev/cuelang.org/go@v0.4.0/pkg/list

**Reply by jlongtine** [reply]
@myitcv Thanks! Yeah, I use those for other things... but getting the unique elements of a list isn't a builtin (yet?).

**Reply by myitcv** [reply]
> getting the unique elements of a list isn't a builtin

No indeed! Was just flagging the existence of that package for other list related tasks. Your answer is spot on!

**Reply by verdverm** [reply]
There seem to be two types of "Unique Elems" functions to me.

1. Unique by key, which is the method suggested by @jlongtine which will error on duplicate keys having different values (https://cuelang.org/play/?id=JF9bFBe3KP5#cue@export@cue)
2. Unique by value, where there is not a well defined unique key to make the determination, which can be implemented using the `list` package https://cuelang.org/play/?id=OjvB30G3YPA#cue@export@cue  
(`uniq: [ for i, x in elems if !list.Contains(list.Drop(elems, i+1), x) { x }]`)

Both are likely common enough use cases and likely be more performant with a Go implementation

**Reply by jlongtine** [reply]
> No indeed! Was just flagging the existence of that package for other list related tasks. Your answer is spot on!

Thanks!
